content/en/blog/_posts/2024-11-21-API-Streaming/2024-11-21-API-Streaming.md

```diff
@@ -0,0 +1,91 @@
+Depending on how the apiserver is run on the node, it might either be killed through OOM by the kernel when exceeding the configured memory limits during these uncontrolled spikes, or if limits are not configured it might have even worse impact on the control plane node.
+And worst, after dying of the first apiserver, the same requests will likely hit another control plane node in an HA setup with probably the same impact. 
+Potentially a situation that is hard to diagnose and hard to recover from.
+
+## Streaming LIST requests
+
+Today, we're excited to announce a major improvement. 
+With the graduation of the WatchList feature to beta in Kubernetes 1.32, client-go users can opt-in (after explicitly enabling WatchListClient feature gate) 
+to streaming lists by switching from LIST to (a special kind of) WATCH requests. 
+
+WATCH requests are served from the watch-cache, an in-memory cache designed to improve scalability of read operations. 
+By streaming each item individually instead of returning the entire list, this method maintains constant memory overhead. 
+The server is bound by the maximum allowed size of an object in etcd plus a few additional allocations. 
+This approach drastically reduces the temporary memory usage compared to traditional LIST requests, ensuring a more efficient and stable system, 
+especially in clusters with a large number of objects of a given type or large average object sizes where despite paging memory consumption used to be high.
+
+Building on the insight gained from the synthetic test (see the synthetic test section for more details), we developed an automated performance test to systematically evaluate the impact of the WatchList feature. 
+This test replicates the same scenario, generating a large number of secrets with a large payload and scaling the number of informers to simulate heavy LIST request patterns. 
+The automated test is executed periodically to monitor memory usage of the server with the feature enabled and disabled.
+
+The results showed significant improvements with the WatchList feature enabled. 
+With the feature turned on, the kube-apiserver’s memory consumption stabilized at approximately **2 GB**. 
+In contrast, with the feature disabled, memory usage increased to approximately **20GB**, a **10x** increase! 
+These results confirm the effectiveness of the new streaming API, which reduces the temporary memory footprint.
+
+## Enabling API Streaming for your component
+
+Upgrade to Kubernetes 1.32. Make sure your cluster uses etcd in version 3.4.31+ or 3.5.13+. 
+Enable WatchListClient for client-go. For details on enabling the feature gate in client-go, please visit https://kubernetes.io/blog/2024/08/12/feature-gates-in-client-go. 
+
+## What's Next?
+In Kubernetes 1.32, the feature is enabled in kube-controller-manager by default despite its beta state. 
+This will be expanded to other core components like kube-scheduler or kubelet eventually, when it is promoted to GA the latest. 
```

Comment:
How about:
```suggestion
This will eventually be expanded to other core components like kube-scheduler or kubelet; once the
client-go feature becomes generally available, all the API clients within core Kubernetes will use
watch lists, if not earlier.
```